src/peer_select.cc

```diff
@@ -111,7 +231,7 @@ PeerSelector::~PeerSelector()
         debugs(44, 3, entry->url());
 
         if (entry->ping_status == PING_WAITING)
-            eventDelete(HandlePingTimeout, this);
+            cancelPingTimeoutMonitoring();
```

Comment:
This cancelation logic looks dangerous (at best).

1. The monitoring service does not use cbdata to protect the service from PeerSelector disappearance because the service relies on this PeerSelector destructor to unregister its object. I think this design, in itself, is permissible because, even with cbdata protections, such a deregistration would required to prevent accumulation of stale monitoring reservations.

2. The current destructor implementation makes derigestration conditional on the ping status. That condition feels very dangerous (and possibly even already wrong -- I have not checked) to me. Sooner or later, the `entry` might become nil or the `ping_status` might change without monitor knowledge -- those two members are not under monitor control. I think deregistration should be conditional on the presence of registration.

To avoid awkward out-of-scope registration position checks in PeerSelector, I suggest calling cancelPingTimeoutMonitoring() unconditionally from the PeerSelector destructor. Let the monitor's forget() method check whether the registration position is set (and do nothing if it is npos); there is already forget() code for that; we just need to make the corresponding source code comment less specific).

If you agree with the solution sketched above, please implement it while removing unnecessary asserts from cancelPingTimeoutMonitoring().

After the above changes, we do not have to keep the cancelPingTimeoutMonitoring() method at all, but I would probably keep it because it encapsulates the (currently very simple) monitor search logic and helps make the code more self-documenting. Combined with the other change request, the method will use pingMonitor().

Reply:
> t calling cancelPingTimeoutMonitoring() unconditionally from the PeerSelector destructor

Done.